engine/select_optimal_members.py

```diff
@@ -0,0 +1,305 @@
+                    f,
+                )
+
+                valid_members_np = np.array(valid_members)
+                indices = [i for i, value in enumerate(passed) if value == 0]
+                failed = valid_members_np[indices]
+                # Increase weights for members which failed
+                weights[failed - 1] += 1 / total_member_num
+                # weights needs to sum up to 1 for np.random.choice
+                weights = weights / sum(weights)
+
+                vars.extend(new_vars)
+
+                if (mem_num == max_member_num) and (f == max_factor):
+                    duplicates = {item: count for item, count in Counter(vars).items()}
+                    sorted_duplicates = dict(
+                        sorted(duplicates.items(), key=lambda x: x[1], reverse=True)
+                    )
+                # The following is to save computing time
+                elif i < 33:
+                    if max_passed < sum(passed):
+                        max_passed = sum(passed)
+                    # The more combs were tested
+                    # the higher should the success rate be to continue
+                    tested_stats = len(valid_members)
+                    if max_passed < i * 0.03 * tested_stats:
+                        break
+
+                if sum(passed) == len(valid_members):
+                    return random_numbers, f
+        # If factore needs to be increased, test only with max_members
+        min_member_num = max_member_num
+
+    max_count = max(sorted_duplicates.values())
+    most_common_vars = [
+        item for item, count in sorted_duplicates.items() if count == max_count
+    ]
+    logger.info(
+        (
+            "ERROR: Could not find {} random members, which pass for all stat files. "
+            + "The most sensitive variable(s) is/are {}, which failed for {} out of {} "
+            + "random selections. Consider removing this/those variable(s) from the "
+            + "experiment and run again."
+        ).format(max_member_num, most_common_vars, max_count, i)
+    )
+    exit(1)
+
+
+def test_selection(
+    stats_file_name, tolerance_file_name, total_member_num, member_type, factor
+):
+    if type(total_member_num) is int:
+        members = [i for i in range(1, total_member_num + 1)]
+    else:
+        members = total_member_num
+        total_member_num = len(members)
+
+    passed = [0] * total_member_num
+
+    input_file_ref = stats_file_name.format(member_id="ref")
+    df_tol = parse_probtest_csv(tolerance_file_name, index_col=[0, 1])
+    df_tol *= factor
+    df_ref = parse_probtest_csv(input_file_ref, index_col=[0, 1, 2])
+
+    vars = []
+    i = 0
+    for m_num in members:
+        m_id = str(m_num) if not member_type else member_type + "_" + str(m_num)
+
+        df_cur = parse_probtest_csv(
+            stats_file_name.format(member_id=m_id), index_col=[0, 1, 2]
+        )
+
+        # check if variables are available in reference file
```

Comment:
I added the function `tolerance_test` to the `check.py` script: https://github.com/MeteoSwiss/probtest/pull/26/commits/7f8774ec3e4eeb32f56fe261a14bd2284b6ebbf8

What do you think?

Reply:
I think it is great. I only have two minor suggestions :upside_down_face: I would locate the function if possible somewhere in utils (maybe dataframe_ops.py or utils.py?). The I would rename it. I like when functions are named by verbs or start with a verb, so I would go with something like `test_stats_files_with_tolerances` or `test_with_tolerances`.

Reply:
The current function `tolerance_test` is using the function `check_intersection` and `check_variable`, all in the `check.py` script. Would you recommend to move all three functions then?

Reply:
Yes, I think it would make sense to move them as well and I think dataframe_ops is probably then the right location. I think this would make the code structure a bit more cleaner.

Reply:
I moved all three functions and renamed `tolerance_test` to `test_stats_file_with_tolerances`

Reply:
Thanks.